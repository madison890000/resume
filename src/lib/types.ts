export enum Degree {
    Bachelor = 'Bachelor',
    Master = 'Master',
    PhD = 'PhD',
}

export type Gender = 'male' | 'female';


export type Achievement = {
    text: string;
    categories: string[];
};
export type SolutionsOfHowToImplement = string;

export enum JobPositionLevel {
    Staff = 'Staff',
    Senior = 'Senior',
    Middle = '',
    Junior = 'Junior',
}

export type JobPosition =
    'Software Engineer'
    | 'Front-End Engineer'
    | 'Back-End Engineer'
    | 'Full-Stack Engineer';

export type JobResponsibility = string;
export type FEUILib = 'Material UI' | 'antd';
export type ReactFramework = 'NextJs' | 'Create React App' | 'umi';
export type Framework = 'React' | 'Vue' | 'Angular' | 'jQuery';
export type CodeLanguage = 'JAVA' | 'C#' | 'JavaScript';
export type OtherSkillLabels = 'NodeJs' | 'HTML5' | 'CSS3' | 'Webpack';
export type WayOfDevelopment = 'Agile' | 'Waterfall';

export type Keyword =
    Framework
    & CodeLanguage
    & OtherSkillLabels
    & FEUILib
    & ReactFramework
    & WayOfDevelopment;

export enum SkillLevel {
    understand = 'understand',
    familiar = 'familiar',
    proficient = 'proficient',
}

export enum Importance {
    important,
    normal,
    less,
}
export enum CompanyType {
    Enterprise,
    Startup,
}
