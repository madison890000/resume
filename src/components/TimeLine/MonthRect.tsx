import React from 'react';

const MonthRect = ({
                       gap = 2,
                       color,
                       index,
                       width = 20,
                   }: {
    index: number;
    width?: number
    color?: string;
    gap?: number;
}) => {
    return (
        <rect width={width} height="10" x={index * (width + gap)} y="5"
              fill={color}></rect>
    )
}

export default MonthRect
