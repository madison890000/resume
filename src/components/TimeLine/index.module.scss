.timelineContainer {
  width: 100%;
}

.yearContainer {
  display: flex;
}

.month-rect {
  width: 10px;
  height: 10px;
}

.jobName {
  font-size: 12px;
  width: 80px;
}
