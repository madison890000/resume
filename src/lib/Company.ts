import { CompanyType } from './types';

interface ICompany {
    name: string;
    location?: string;
    type?: CompanyType;
}

export default class Company {
    public name: string;
    public type?: CompanyType;
    public location?: string;

    constructor({ name, location, type }: ICompany) {
        this.name = name;
        this.location = location;
        this.type = type;
    }

}
