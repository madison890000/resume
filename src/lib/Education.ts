import { Degree } from './types';


interface IEducation {
    college: string;
    major: string;
    degree: Degree;
    start: Date;
    end: Date;
}

export default class Education {
    public college: string;
    public major: string;
    public degree: Degree;
    public start: Date;
    public end: Date;

    constructor({ college,major, degree, start, end }: IEducation) {
        this.major = major;
        this.college = college;
        this.degree = degree;
        this.start = start;
        this.end = end;
    }
}
