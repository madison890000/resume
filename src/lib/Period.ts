import Company from './Company';
import { v4 } from 'uuid';
import {
    Achievement,
    JobPosition,
    JobPositionLevel,
    Keyword,
    SolutionsOfHowToImplement
} from './types';

interface IPeriod {
    start: Date;
    end: Date;
    keywords: Keyword[];
    company: Company;
    jobPositionLevel: JobPositionLevel;
    jobPosition: JobPosition;
    achievements: Achievement[];
    jobSummaries: string[];
    solutionsOfHowToImplement: SolutionsOfHowToImplement[];
}

export default class Period {
    readonly id: string;
    public start: Date;
    public end: Date;
    public keywords: Keyword[];
    public achievements: Achievement[];
    public jobSummaries: string[];
    public company: Company;
    public jobPositionLevel: JobPositionLevel;
    public jobPosition: JobPosition;
    public solutionsOfHowToImplement: SolutionsOfHowToImplement[];

    constructor({solutionsOfHowToImplement, company, jobSummaries, jobPosition, jobPositionLevel, achievements, keywords, start, end }: IPeriod) {
        this.id = v4();
        this.company = company;
        this.start = start;
        this.end = end;
        this.achievements = achievements;
        this.jobSummaries = jobSummaries;
        this.keywords = keywords;
        this.jobPositionLevel = jobPositionLevel;
        this.jobPosition = jobPosition;
        this.solutionsOfHowToImplement = solutionsOfHowToImplement;
    }

}
