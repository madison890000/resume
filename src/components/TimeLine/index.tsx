import React, { useEffect, useState } from 'react';
import styles from './index.module.scss';
import Period from '../../lib/Period';
import { findPeriodByDate, getMonthCountInStartAndEnd } from '../../utils/date';
import MonthRect from './MonthRect';


const COLORS = [
    '#FF6666',
    '#FFFF66',
    '#66FFFF',
    '#6666FF',
    '#B266FF',
    '#FF66B2',
]

const PAGE_PADDING = 24;

const TimeLine = ({ periods }: { periods: Period[] }) => {
    const start = periods?.[0]?.start;
    const end = periods?.[periods?.length - 1]?.end;
    const [width, setWidth] = useState<number>(0);
    useEffect(() => {
        const { innerWidth } = window;
        setWidth(innerWidth - PAGE_PADDING - 40)
    }, [])
    const rectGap = 1;
    const totalRects = getMonthCountInStartAndEnd(start, end) + 1;
    const rectFullWidth = width / totalRects;
    const rectWidth = rectFullWidth - rectGap;
    const periodColors: { [key: string]: string } = {};
    periods?.forEach((p, index) => {
        periodColors[p.id] = COLORS[index];
    })
    let rectArray = new Array(totalRects).fill(0)
        ?.map((_, index) => {
            const year = start.getFullYear() + Math.floor((start.getMonth() + index) / 12);
            const month = (start.getMonth() + index) % 12;
            const period = findPeriodByDate(new Date(Number(year), month, 1), periods)
            return {
                index: index,
                period: period,
                color: period ? periodColors[period?.id] : '',
            }
        })
    return (
        <div className={styles.timelineContainer}>
            <div>
                <svg width={width} height="20" style={{ backgroundColor: 'E0E0E0' }}>
                    <g>
                        {rectArray?.map((e, index) => (
                            <MonthRect
                                width={rectWidth}
                                gap={rectGap}
                                color={e?.color}
                                index={e?.index}
                            />
                        ))}
                    </g>
                </svg>
            </div>
            <div style={{
                height: 48
            }}>
                {periods?.map((p, index) => {
                    const translateX = (getMonthCountInStartAndEnd(start, p?.start) + 1) * rectFullWidth;
                    return (
                        <span
                            style={{
                                position: 'absolute',
                                transform: `translateX(${translateX}px)`
                            }}
                            className={styles.jobName}
                        >
                        {p?.jobPositionLevel} {p?.jobPosition}
                    </span>
                    )
                })}
            </div>
        </div>
    )
}

export default TimeLine
