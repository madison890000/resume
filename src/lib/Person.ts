import Education from './Education';
import Period from './Period';
import { Gender } from './types';
import Skill from './Skill';


interface IPerson {
    firstName: string;
    lastName: string;
    email: string;
    birthDay: Date;
    gender: Gender;
    cellphone: string;
    country: string;
    description: string[];
    educations?: Education[];
    periods?: Period[];
}


export default class Person {
    public firstName: string;
    public lastName: string;
    public email: string;
    public gender: Gender;
    public cellphone: string;
    public country: string;
    public birthDay: Date;
    public educations: Education[];
    public periods: Period[];
    public skills: Skill[];
    public description: string[];

    constructor({ firstName,description, birthDay, lastName, email, cellphone, country, gender }: IPerson) {
        this.educations = [];
        this.periods = [];
        this.firstName = firstName;
        this.lastName = lastName;
        this.email = email;
        this.cellphone = cellphone;
        this.birthDay = birthDay;
        this.country = country;
        this.gender = gender;
        this.skills = [];
        this.description = description;
    }

    addEducations(educations: Education[]) {
        this.educations = this.educations.concat(educations);
    }

    addPeriods(periods: Period[]) {
        this.periods = this.periods.concat(periods);
    }

    addSkills(skills: Skill[]) {
        this.skills = this.skills.concat(skills);
    }

    get reversedPeriods() {
        return [...this.periods].reverse()
    }

}
