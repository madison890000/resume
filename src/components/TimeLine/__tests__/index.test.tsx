import React from 'react';
import TimeLine from '../index';
import renderer from 'react-test-renderer';
import { Achievement, JobPositionLevel, Keyword } from '../../../lib/types';
import Company from '../../../lib/Company';
import Period from '../../../lib/Period';

describe('renders TimeLine ', () => {
    const testCompanyName = 'test company name'
    const testCompany = new Company({
        name: testCompanyName,
    })
    const start = new Date('2012-01-01');
    const end = new Date('2013-01-01');
    const keywords = ['React'] as Keyword[];
    const jobPositionLevel = JobPositionLevel.Senior;
    const jobPosition = 'Software Engineer';
    const achievements = [] as Achievement[];
    const jobSummaries = [] as string[];
    const testPeriod = new Period({
        solutionsOfHowToImplement: [],
        company: testCompany,
        jobSummaries,
        jobPositionLevel,
        jobPosition,
        achievements,
        start,
        end,
        keywords
    });
    test('TimeLine one period', () => {
        const TimeLineOne = renderer.create(
            <TimeLine
                periods={[testPeriod]}
            />
        )
            .toJSON();
        expect(TimeLineOne)
            .toMatchSnapshot();
    })
});
