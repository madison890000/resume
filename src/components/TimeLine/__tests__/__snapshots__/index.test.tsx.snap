// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TimeLine  TimeLine one period 1`] = `
<div
  className="timelineContainer"
>
  <div>
    <svg
      height="20"
      style={
        Object {
          "backgroundColor": "E0E0E0",
        }
      }
      width={0}
    >
      <g>
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={-1}
          x={0}
          y="5"
        />
      </g>
    </svg>
  </div>
  <div
    style={
      Object {
        "height": 48,
      }
    }
  >
    <span
      className="jobName"
      style={
        Object {
          "position": "absolute",
          "transform": "translateX(0px)",
        }
      }
    >
      Senior
       
      Software Engineer
    </span>
  </div>
</div>
`;
